lib/avromatic/model/attributes.rb

```diff
@@ -44,8 +44,22 @@ def required?
 
         def coerce(input)
           type.coerce(input)
+        rescue Avromatic::Model::UnknownAttributeError => e
+          raise Avromatic::Model::CoercionError.new("Value for #{owner.name}##{name} could not be coerced to a #{type.name} " \
+            "because the following unexpected attributes were provided: #{e.attributes.join(', ')}. " \
+            "Provided argument: #{input.inspect}")
         rescue StandardError
-          raise Avromatic::Model::CoercionError.new("Could not coerce '#{input.inspect}' to #{type.name} when setting #{owner.name}##{name}")
+          if type.input_classes && type.input_classes.none? { |input_class| input.is_a?(input_class) }
+            raise Avromatic::Model::CoercionError.new("Value for #{owner.name}##{name} could not be coerced to a #{type.name} " \
+              "because a #{input.class.name} was provided but expected a #{type.input_classes.map(&:name).to_sentence(two_words_connector: ' or ', last_word_connector: ', or ')}. " \
+              "Provided argument: #{input.inspect}")
+          elsif input.is_a?(Hash) && type.is_a?(Avromatic::Model::Types::UnionType)
+            raise Avromatic::Model::CoercionError.new("Value for #{owner.name}##{name} could not be coerced to a #{type.name} " \
+              "because no union member type has all of the provided attributes: #{input.inspect}")
```

Comment:
If it's straightforward to print the union member types' attributes, that would be helpful for comparison with the provided attributes.

Reply:
We talked about this offline but I'm going to punt on doing this as part of this PR.